File: app/src/main/java/com/boardgamegeek/provider/GamesProvider.java
package com.boardgamegeek.provider;

import android.net.Uri;
import android.text.TextUtils;

import com.boardgamegeek.provider.BggContract.Games;
import com.boardgamegeek.provider.BggContract.Plays;
import com.boardgamegeek.provider.BggDatabase.Tables;
import com.boardgamegeek.util.SelectionBuilder;

public class GamesProvider extends BasicProvider {

	@Override
	protected String getDefaultSortOrder() {
		return Games.DEFAULT_SORT;
	}

	@Override
	protected String getInsertedIdColumn() {
		return Games.GAME_ID;
	}

	@Override
	protected String getPath() {
		return BggContract.PATH_GAMES;
	}

	@Override
	protected String getTable() {
		return Tables.GAMES;
	}

	@Override
	protected String getType(Uri uri) {
		return Games.CONTENT_TYPE;
	}

	@Override
	protected SelectionBuilder buildExpandedSelection(Uri uri) {
		SelectionBuilder builder = new SelectionBuilder();
		if (BggContract.FRAGMENT_PLAYS.equals(uri.getFragment())) {
			builder
				.table(Tables.GAMES_JOIN_PLAYS)
				.mapIfNull(Games.GAME_RANK, String.valueOf(Integer.MAX_VALUE))
				.mapAsSum(Plays.SUM_QUANTITY, Plays.QUANTITY, Tables.PLAYS)
				.mapAsMax(Plays.MAX_DATE, Plays.DATE);
		} else {
			builder
				.table(Tables.GAMES_JOIN_COLLECTION)
				.mapToTable(Games.GAME_ID, Tables.GAMES);
		}
		builder.mapToTable(Games.UPDATED, Tables.GAMES);


		String groupBy = uri.getQueryParameter(BggContract.QUERY_KEY_GROUP_BY);
		if (!TextUtils.isEmpty(groupBy)) {
			builder.groupBy(groupBy);
		} else {
			builder.groupBy(Games.GAME_ID);
		}
		return builder;
	}
}

File: app/src/main/java/com/boardgamegeek/provider/GamesProvider.kt
package com.boardgamegeek.provider

import android.net.Uri
import com.boardgamegeek.provider.BggContract.Companion.FRAGMENT_PLAYS
import com.boardgamegeek.provider.BggContract.Companion.PATH_GAMES
import com.boardgamegeek.provider.BggContract.Games
import com.boardgamegeek.provider.BggContract.Plays
import com.boardgamegeek.provider.BggDatabase.Tables

class GamesProvider : BasicProvider() {
    override fun getType(uri: Uri) = Games.CONTENT_TYPE

    override val path = PATH_GAMES

    override val table = Tables.GAMES

    override val defaultSortOrder = Games.DEFAULT_SORT

    override val insertedIdColumn = Games.Columns.GAME_ID

    override fun buildExpandedSelection(uri: Uri): SelectionBuilder {
        val builder = SelectionBuilder()
        if (FRAGMENT_PLAYS == uri.fragment) {
            builder
                .table(Tables.GAMES_JOIN_PLAYS)
                .mapAsSum(Plays.Columns.SUM_QUANTITY, Plays.Columns.QUANTITY, Tables.PLAYS)
                .mapAsMax(Plays.Columns.MAX_DATE, Plays.Columns.DATE)
        } else {
            builder
                .table(Tables.GAMES_JOIN_COLLECTION)
                .mapToTable(Games.Columns.GAME_ID, Tables.GAMES)
        }
        return builder
            .mapToTable(Games.Columns.UPDATED, Tables.GAMES)
            .groupBy(Games.Columns.GAME_ID)
    }
}
